informalsystems/ibc-rs
# Fix for chain impersonation problem. Fix for incomplete `channel create`

Closes: #1038
Closes: #1064

For testing instructions for issue 1064, see the description within that issue for more details.
The basic command that is undergoing changes here is `hermes create channel`.



## Testing for 1038

You will need two configuration files for Hermes.

1. save [this](https://gist.github.com/adizere/5169bea5176819954c9767ffd7e3fa42#file-config-toml) into `~/.hermes/config.toml`
2. save [this](https://gist.github.com/adizere/5169bea5176819954c9767ffd7e3fa42#file-config_fake-toml) into `~/.hermes/config_fake.toml`

You will also need to use [this](https://gist.github.com/adizere/5169bea5176819954c9767ffd7e3fa42#file-gm-toml) for `gm`. Make sure that the `binary` option in that file points to the correct path to your hermes binary.


1. Make sure no gaia instance is running, i.e., `killall gaiad` and `gm stop`
    - if some `gm` chains were running before, `gm reset` them to clear their state
2. do `./scripts/dev-env ~/.hermes/config.toml ibc-0 ibc-1` and wait for it finish
    - the chain `ibc-1` on grpc port `9091` will be our real chain where the ft-transfer should arrive
    - the config for this chain is consistent with that in file `config.toml`
3.  `gm start` && `gm hermes keys`
    - this starts another chain, also called `ibc-1`, but this gaia is running on a different grpc port `27041`
    - this will be our `ibc-1` that is fake 
    - the config for this chain is consistent with that in file `config_fake.toml`
4. `hermes create channel ibc-0 ibc-1 --port-a transfer --port-b transfer`
    - creates a channel: `ibc-0:channel-0` ----> real `ibc-1:channel-1`
5. `hermes -c ~/.hermes/config_fake.toml create channel ibc-0 ibc-1 --port-a transfer --port-b transfer`
    - creates a channel: `ibc-0:channel-1` ----> fake `ibc-1:channel-1`
    - the complete picture of the state of the three chains is as follows
```
ibc-1 (real)                                ibc-0
     (channel-0, connection-0, client-0) <-->   (channel-0, connection-0, client-0)

ibc-1 (fake)                                ibc-0
     (channel-0, connection-0, client-0) <-->   (channel-1, connection-1, client-1)
```

6. now submit an ft-transfer on `ibc-0`, on end `ibc-0:channel-1`, destined to arrive to the fake `ibc-1` chain, 
    - `hermes -c ~/.hermes/config_fake.toml tx raw ft-transfer ibc-1 ibc-0 transfer channel-1 9999 -o 1000`
    - this behavior is correct, so far

7. try to submit an ft-transfer on `ibc-0`, on end `ibc-0:channel-1`, destined to arrive to the real `ibc-1` chain
    - we omit the `-c` option to use the config of the real `ibc-1` chain
    - note that we use `ibc-0:channel-1` to submit this transfer: this specific channel end on `ibc-0` corresponds to a counterparty channel end `ibc-1:channel-0`, which in turn has counterparty `ibc-0:channel-0`; in other words, this command should fail, because the counterparty on the destination chain (namely, `ibc-0:channel-0`) does not match the source channel (`ibc-0:channel-1`):
    - `hermes tx raw ft-transfer ibc-1 ibc-0 transfer channel-1 9999 -o 1000`
    - with the changes in this PR, the following message is output:

> Error: socket transfer/channel-0 on chain ibc-1 expected to have counterparty transfer/channel-1 (but instead has transfer/channel-0)


______

For contributor use:

- [x] Updated the __Unreleased__ section of [CHANGELOG.md](https://github.com/informalsystems/ibc-rs/blob/master/CHANGELOG.md) with the issue.
